How to disable `whereToFetchOnline` and `changeButtonProvider` ?

```
UnhandledPromiseRejectionWarning: TypeError: filterWrapper is not iterable
    at Object.exports.parseFilters (D:\Tools\vlc-rpc\node_modules\ytsr\lib\utils.js:23:29)
    at module.exports (D:\Tools\vlc-rpc\node_modules\ytsr\lib\main.js:60:24)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async module.exports (D:\Tools\vlc-rpc\src\rpc\format.js:124:22)
    at async D:\Tools\vlc-rpc\src\rpc\client.js:24:26
```
It shows a lot of these when changing the track.

`D:\Tools\vlc-rpc>start.bat`
```
Did not detect node_modules directory; Installing modules...
Modules installed.
Starting script...

> vlc-rpc@1.1.9 start
> node ./src/app.js

node:internal/modules/cjs/loader:1080
  throw err;
  ^

Error: Cannot find module 'vlc.js'
Require stack:
- D:\Tools\vlc-rpc\src\helpers\lager.js
- D:\Tools\vlc-rpc\src\app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (D:\Tools\vlc-rpc\src\helpers\lager.js:5:13)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1143:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'D:\\Tools\\vlc-rpc\\src\\helpers\\lager.js',
    'D:\\Tools\\vlc-rpc\\src\\app.js'
  ]
}

Node.js v18.16.1
```
`D:\Tools\vlc-rpc>npm i --production`
```
npm WARN config production Use `--omit=dev` instead.
npm ERR! code 1
npm ERR! git dep preparation failed
npm ERR! command C:\Program Files\nodejs\node.exe C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\bin\npm-cli.js install --force --cache=C:\Users\nguye\AppData\Local\npm-cache --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm ERR! > register-scheme@0.0.2 install
npm ERR! > node-gyp rebuild
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@18.16.1 | win32 | x64
npm ERR! gyp info find Python using Python version 3.10.8 found at "C:\msys64\mingw64\bin\python3.exe"
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! find VS msvs_version not set from command line or npm config
npm ERR! gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
npm ERR! gyp ERR! find VS could not use PowerShell to find Visual Studio 2017 or newer, try re-running with '--loglevel silly' for more details
npm ERR! gyp ERR! find VS looking for Visual Studio 2015
npm ERR! gyp ERR! find VS - not found
npm ERR! gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS You need to install the latest version of Visual Studio
npm ERR! gyp ERR! find VS including the "Desktop development with C++" workload.
npm ERR! gyp ERR! find VS For more information consult the documentation at:
npm ERR! gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! configure error
npm ERR! gyp ERR! stack Error: Could not find any Visual Studio installation to use
npm ERR! gyp ERR! stack     at VisualStudioFinder.fail (C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:122:47)
npm ERR! gyp ERR! stack     at C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:75:16
npm ERR! gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:380:14)
npm ERR! gyp ERR! stack     at C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:71:14
npm ERR! gyp ERR! stack     at C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:401:16
npm ERR! gyp ERR! stack     at C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\util.js:54:7
npm ERR! gyp ERR! stack     at C:\Users\nguye\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\util.js:33:16
npm ERR! gyp ERR! stack     at ChildProcess.exithandler (node:child_process:427:5)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
npm ERR! gyp ERR! stack     at maybeClose (node:internal/child_process:1091:16)
npm ERR! gyp ERR! System Windows_NT 10.0.19045
npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\nguye\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd C:\Users\nguye\AppData\Local\npm-cache\_cacache\tmp\git-clonenUtmyq
npm ERR! gyp ERR! node -v v18.16.1
npm ERR! gyp ERR! node-gyp -v v9.4.0
npm ERR! gyp ERR! not ok
npm ERR! npm ERR! code 1
npm ERR! npm ERR! path C:\Users\nguye\AppData\Local\npm-cache\_cacache\tmp\git-clonenUtmyq
npm ERR! npm ERR! command failed
npm ERR! npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node-gyp rebuild
npm ERR!
npm ERR! npm ERR! A complete log of this run can be found in: C:\Users\nguye\AppData\Local\npm-cache\_logs\2023-07-25T02_55_25_542Z-debug-0.log

npm ERR! A complete log of this run can be found in: C:\Users\nguye\AppData\Local\npm-cache\_logs\2023-07-25T02_55_23_365Z-debug-0.log
```
@sdasda7777

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -51,9 +51,24 @@ const fetchers = {
     };
   },
   "youtube": async (artist, album, song) => {
-    // This just crashes most of the time, no clue why
-    const result = await yt(encodeURI(`${artist} ${song}`), { limit: 1 });
-    if (result.items.length == 0)
+    try
+    {
+      // This just crashes most of the time, no clue why, hence the try-catch-finally
+      const result = await yt(encodeURI(`${artist} ${song}`), { limit: 1 });
+      if (result.items.length > 0)
+      {
+        return {
+          fetchedFrom: "Youtube",
+          artworkUrl: undefined,
+          joinUrl: result.items[0].url
+        };
+      }
+    }
+    catch(err)
+    {
+      console.log(err);
+    }
+    finally
     {
       return {
         fetchedFrom: undefined,
@@ -61,15 +76,73 @@ const fetchers = {
         joinUrl: undefined
       };
     }
-    else
+  }
+}
+
+// Fetches both artwork and join URL if possible,
+// Returns object containing artworkUrl, artworkFrom, joinUrl and joinFrom
+async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, artist, album, song)
+{
+  let artworkUrl, artworkFrom, joinUrl, joinFrom;
+  let results = [];
+
+  // First try fetching artwork URL using preferred provider
+  if (preferredArtworkProvider in fetchers)
+  {
+    results[preferredArtworkProvider] = await fetchers[preferredArtworkProvider](artist, album, song);
+    if(results[preferredArtworkProvider].artworkUrl)
     {
-      return {
-        fetchedFrom: "Youtube",
-        artworkUrl: undefined,
-        joinUrl: result.items[0].url
-      };
+      artworkUrl = results[preferredArtworkProvider].artworkUrl;
+      artworkFrom = results[preferredArtworkProvider].fetchedFrom;
     }
   }
+
+  // Next try fetching join URL
+  if (preferredArtworkProvider !== preferredJoinProvider
+      && preferredJoinProvider in fetchers)
+  {
+    results[preferredJoinProvider] = await fetchers[preferredJoinProvider](artist, album, song);
+    if (results[preferredJoinProvider].joinUrl)
+    {
+      joinUrl = results[preferredJoinProvider].joinUrl;
+      joinFrom = results[preferredJoinProvider].fetchedFrom;
+    }
+  }
+
+  // Try using preferred join provider as a backup artwork provider and vice versa
+  if (!artworkUrl && results[preferredJoinProvider].artworkUrl)
+  {
+    artworkUrl = results[preferredJoinProvider].artworkUrl;
+    artworkFrom = results[preferredJoinProvider].fetchedFrom;
+  }
+  if (!joinUrl && results[preferredArtworkProvider].joinUrl)
+  {
+    joinUrl = results[preferredArtworkProvider].joinUrl;
+    joinFrom = results[preferredArtworkProvider].fetchedFrom;
+  }
+
+  // In case either still isn't set, iterate all other providers
+  let availableProviderNames = Object.keys(fetchers);
+  for (let ii = 0; (!artworkUrl || !joinUrl) && ii < availableProviderNames.length; ++ii)
+  {
+    if (!(availableProviderNames[ii] in results))
+    {
+      results[availableProviderNames[ii]] = await fetchers[availableProviderNames[ii]](artist, album, song);
+      if(!artworkUrl && results[availableProviderNames[ii]].artworkUrl)
+      {
+        artworkUrl = results[availableProviderNames[ii]].artworkUrl;
+        artworkFrom = results[availableProviderNames[ii]].fetchedFrom;
+      }
+      if (!joinUrl && results[availableProviderNames[ii]].joinUrl)
+      {
+        joinUrl = results[availableProviderNames[ii]].joinUrl;
+        joinFrom = results[availableProviderNames[ii]].fetchedFrom;
+      }
+    }
+  }
+
+  // Return selected results
+  return {artworkUrl, artworkFrom, joinUrl, joinFrom};
 }
 
 
@@ -105,12 +178,13 @@ module.exports = async (status) => {
 
   if (meta.title !== undefined && display_artist !== undefined)
   {
-    // Initial fetch for artwork
-    const fetchResult1 = await fetchers[config.rpc.whereToFetchOnline](display_artist, meta.album, meta.title);
-    if(fetchResult1.artworkUrl != undefined)
+    const fetchResult = await combinedFetch(config.rpc.whereToFetchOnline, config.rpc.changeButtonProvider, display_artist, meta.album, meta.title);
+
+    // Set artwork URL if present
+    if(fetchResult.artworkUrl != undefined)
     {
-      artwork = fetchResult1.artworkUrl;
-      fetched = fetchResult1.fetchedFrom;
+      artwork = fetchResult.artworkUrl;
+      fetched = fetchResult.artworkFrom;
     }
     else
     {
@@ -118,24 +192,11 @@ module.exports = async (status) => {
       fetched = "Nowhere";
     }
 
-    // If desired artwork and join URL providers are the same,
-    //  use existing fetch, otherwise fetch the join URL
-    if (config.rpc.whereToFetchOnline === config.rpc.changeButtonProvider)
+    // Set join URL if present
+    if (fetchResult.joinUrl)
     {
-      if (fetchResult1.joinUrl)
-      {
-        joinUrl = fetchResult1.joinUrl;
-        joinLabel = `Listen on ${fetchResult1.fetchedFrom}`;
-      }
-    }
-    else
-    {
-      const fetchResult2 = await fetchers[config.rpc.changeButtonProvider](display_artist, meta.album, meta.title);
-      if(fetchResult2.joinUrl)
-      {
-        joinUrl = fetchResult2.joinUrl;
-        joinLabel = `Listen on ${fetchResult2.fetchedFrom}`;
-      }
+      joinUrl = fetchResult.joinUrl;
+      joinLabel = `Listen on ${fetchResult.joinFrom}`;
     }
   }
   else

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -12,115 +12,107 @@ const yt = require("ytsr");
 
 
 
-// These functions provide uniform inteface for simple access to the APIs
-// They take artist album and song as the arguments and return object containing
+// These functions, 'fetchers', provide uniform inteface for simple access to the APIs
+// They take artist album and song as the arguments and on success return object containing
 //   fetchedFrom, artworkUrl and joinUrl (or undefined where not applicable).
+// In order to be as simple as possible, the functions may throw exceptions or not return anything.
+//   They are expected to be called through the `fetchSafely()` wrapper.
 const fetchers = {
   "apple": async (metadata) => {
-    if ((!metadata.albumartist && !metadata.artist) || !metadata.title)
+    if ((metadata.albumartist || metadata.artist) && metadata.title)
     {
-      return {};
-    }
-    const params = {
-      media: "music",
-      term: `${metadata.albumartist ? metadata.albumartist : metadata.artist} ${metadata.title}`,
-    };
-    const result = await axios.get("https://itunes.apple.com/search", {
-      params,
-    });
-    if (result.data.resultCount == 0 || result.data.results[0] === undefined)
-    {
-      return {};
-    }
-    else
-    {
-      return {
-        fetchedFrom: "Apple Music",
-        artworkUrl: result.data.results[0].artworkUrl100,
-        joinUrl: result.data.results[0].trackViewUrl
-      };
+      const result = await axios.get("https://itunes.apple.com/search", {
+        params: {
+          media: "music",
+          term: `${metadata.albumartist ? metadata.albumartist : metadata.artist} ${metadata.title}`,
+        }
+      });
+      if (result.data.resultCount > 0 && result.data.results[0] !== undefined)
+      {
+        returnValue = {
+          fetchedFrom: "Apple Music",
+          artworkUrl: result.data.results[0].artworkUrl100,
+          joinUrl: result.data.results[0].trackViewUrl
+        };
+      }
     }
   },
   "spotify": async (metadata) => {
-    let returnValue;
-    try
+    if ((metadata.albumartist || metadata.artist) && metadata.album)
     {
-      if ((!metadata.albumartist && !metadata.artist) || !metadata.album)
-      {
-        return {};
-      }
-
-      let result = await albumArt(metadata.albumartist ? metadata.albumartist : metadata.artist,
-                                  {album: metadata.album}).then((data) => data);
+      const result = await albumArt(metadata.albumartist ? metadata.albumartist : metadata.artist,
+                                    {album: metadata.album}).then((data) => data);
       // TODO: how is failure indicated in the albumArt library?
-      returnValue = {
+      return {
         fetchedFrom: "Spotify",
         artworkUrl: result,
         joinUrl: undefined
       }
     }
-    catch(err)
-    {
-      console.log(err);
-    }
-    finally
-    {
-      return (returnValue ? returnValue : {});
-    }
   },
   "youtube": async (metadata) => {
-    let returnValue;
-    try
+    const result = await yt(`${metadata.albumartist ? metadata.albumartist : (metadata.artist ? metadata.artist : "")} ${metadata.title ? metadata.title : metadata.filename}`.trim(), { limit: 1 });
+    if (result.items.length > 0)
     {
-      // This just crashes most of the time, no clue why, hence the try-catch-finally
-      const result = await yt(`${metadata.albumartist ? metadata.albumartist : (metadata.artist ? metadata.artist : "")} ${metadata.title ? metadata.title : metadata.filename}`.trim(), { limit: 1 });
-      if (result.items.length > 0)
-      {
-        returnValue = {
-          fetchedFrom: "Youtube",
-          artworkUrl: undefined,
-          joinUrl: result.items[0].url
-        };
-      }
-    }
-    catch(err)
-    {
-      console.log(err);
-    }
-    finally
-    {
-      return (returnValue ? returnValue : {});
+      return {
+        fetchedFrom: "Youtube",
+        artworkUrl: undefined,
+        joinUrl: result.items[0].url
+      };
     }
   }
 }
 
-// Fetches both artwork and join URL if possible,
-// Returns object containing artworkUrl, artworkFrom, joinUrl and joinFrom
+/**
+ * Safe wrapper for calling a fetcher in try-catch block
+ * @param {string} fetcherName name of the fetcher to use
+ * @param {Object} metadata    VLC metadata
+ * @returns {!{fetchedFrom: string, artworkUrl: string, joinUrl: string}}
+ */
+async function fetchSafely(fetcherName, metadata)
+{
+  let returnValue;
+  try
+  {
+    returnValue = await fetchers[fetcherName](metadata);
+  }
+  catch (err)
+  {
+    console.log(err);
+  }
+  finally
+  {
+    return (returnValue ? returnValue : {});
+  }
+}
+
+/**
+ * Fetches both artwork and join URL if possible
+ * @param {string} preferredArtworkProvider name of preferred artwork fetcher
+ * @param {string} preferredJoinProvider    name of preferred join fetcher
+ * @param {Object} metadata                 VLC metadata
+ * @returns {!{artworkUrl: string, artworkFrom: string, joinUrl: string, joinFrom: string}}
+ */
 async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, metadata)
 {
   let artworkUrl, artworkFrom, joinUrl, joinFrom;
   let results = [];
 
   // First try fetching artwork URL using preferred provider
-  if (preferredArtworkProvider in fetchers)
+  results[preferredArtworkProvider] = await fetchSafely(preferredArtworkProvider, metadata);
+  if(results[preferredArtworkProvider].artworkUrl)
   {
-    results[preferredArtworkProvider] = await fetchers[preferredArtworkProvider](metadata);
-    if(results[preferredArtworkProvider].artworkUrl)
-    {
-      artworkUrl = results[preferredArtworkProvider].artworkUrl;
-      artworkFrom = results[preferredArtworkProvider].fetchedFrom;
-    }
+    artworkUrl = results[preferredArtworkProvider].artworkUrl;
+    artworkFrom = results[preferredArtworkProvider].fetchedFrom;
   }
 
   // Next try fetching join URL from preferred provider
-  if (preferredJoinProvider in fetchers
-      && preferredArtworkProvider !== preferredJoinProvider)
+  if (preferredArtworkProvider !== preferredJoinProvider)
   {
-    results[preferredJoinProvider] = await fetchers[preferredJoinProvider](metadata);
+    results[preferredJoinProvider] = await fetchSafely(preferredJoinProvider, metadata);
   }
   // Set it separately, in case both preferred providers are the same
-  if (preferredJoinProvider in results
-      && results[preferredJoinProvider].joinUrl)
+  if (results[preferredJoinProvider].joinUrl)
   {
     joinUrl = results[preferredJoinProvider].joinUrl;
     joinFrom = results[preferredJoinProvider].fetchedFrom;
@@ -144,7 +136,7 @@ async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, me
   {
     if (!(availableProviderNames[ii] in results))
     {
-      results[availableProviderNames[ii]] = await fetchers[availableProviderNames[ii]](metadata);
+      results[availableProviderNames[ii]] = await fetchSafely(availableProviderNames[ii], metadata);
       if(!artworkUrl && results[availableProviderNames[ii]].artworkUrl)
       {
         artworkUrl = results[availableProviderNames[ii]].artworkUrl;

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -13,7 +13,7 @@ const yt = require("ytsr");
 
 
 // These functions, 'fetchers', provide uniform inteface for simple access to the APIs
-// They take artist album and song as the arguments and on success return object containing
+// They take VLC metadata as the argument and on success return object containing
 //   fetchedFrom, artworkUrl and joinUrl (or undefined where not applicable).
 // In order to be as simple as possible, the functions may throw exceptions or not return anything.
 //   They are expected to be called through the `fetchSafely()` wrapper.
@@ -87,7 +87,7 @@ async function fetchSafely(fetcherName, metadata)
 }
 
 /**
- * Fetches both artwork and join URL if possible
+ * Fetches both artwork URL and join URL if possible
  * @param {string} preferredArtworkProvider name of preferred artwork fetcher
  * @param {string} preferredJoinProvider    name of preferred join fetcher
  * @param {Object} metadata                 VLC metadata
@@ -170,19 +170,18 @@ module.exports = async (status) => {
 
   // Fetch artwork and join URLs
   let artwork = config.rpc.largeIcon, fetched = "Nowhere", joinUrl, joinLabel;
-
   {
     const fetchResult
       = await combinedFetch(config.rpc.whereToFetchOnline, config.rpc.changeButtonProvider, meta);
 
-    // Set artwork URL if present
-    if(fetchResult.artworkUrl != undefined)
+    // Use artwork URL if present
+    if(fetchResult.artworkUrl)
     {
       artwork = fetchResult.artworkUrl;
       fetched = fetchResult.artworkFrom;
     }
 
-    // Set join URL if present
+    // Use join URL if present
     if (fetchResult.joinUrl)
     {
       joinUrl = fetchResult.joinUrl;

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -16,21 +16,21 @@ const yt = require("ytsr");
 // They take artist album and song as the arguments and return object containing
 //   fetchedFrom, artworkUrl and joinUrl (or undefined where not applicable).
 const fetchers = {
-  "apple": async (artist, album, song) => {
+  "apple": async (metadata) => {
+    if ((!metadata.albumartist && !metadata.artist) || !metadata.title)
+    {
+      return {};
+    }
     const params = {
       media: "music",
-      term: `${artist} ${song}`,
+      term: `${metadata.albumartist ? metadata.albumartist : metadata.artist} ${metadata.title}`,
     };
     const result = await axios.get("https://itunes.apple.com/search", {
       params,
     });
     if (result.data.resultCount == 0 || result.data.results[0] === undefined)
     {
-      return {
-        fetchedFrom: undefined,
-        artworkUrl: undefined,
-        joinUrl: undefined
-      };
+      return {};
     }
     else
     {
@@ -41,23 +41,42 @@ const fetchers = {
       };
     }
   },
-  "spotify": async (artist, album, song) => {
-    let result = await albumArt(artist, {album}).then((data) => data);
-    // TODO: how is failure indicated in the albumArt library?
-    return {
-      fetchedFrom: "Spotify",
-      artworkUrl: result,
-      joinUrl: undefined
-    };
+  "spotify": async (metadata) => {
+    let returnValue;
+    try
+    {
+      if ((!metadata.albumartist && !metadata.artist) || !metadata.album)
+      {
+        return {};
+      }
+
+      let result = await albumArt(metadata.albumartist ? metadata.albumartist : metadata.artist,
+                                  {album: metadata.album}).then((data) => data);
+      // TODO: how is failure indicated in the albumArt library?
+      returnValue = {
+        fetchedFrom: "Spotify",
+        artworkUrl: result,
+        joinUrl: undefined
+      }
+    }
+    catch(err)
+    {
+      console.log(err);
+    }
+    finally
+    {
+      return (returnValue ? returnValue : {});
+    }
   },
-  "youtube": async (artist, album, song) => {
+  "youtube": async (metadata) => {
+    let returnValue;
     try
     {
       // This just crashes most of the time, no clue why, hence the try-catch-finally
-      const result = await yt(encodeURI(`${artist} ${song}`), { limit: 1 });
+      const result = await yt(`${metadata.albumartist ? metadata.albumartist : (metadata.artist ? metadata.artist : "")} ${metadata.title ? metadata.title : metadata.filename}`.trim(), { limit: 1 });
       if (result.items.length > 0)
       {
-        return {
+        returnValue = {
           fetchedFrom: "Youtube",
           artworkUrl: undefined,
           joinUrl: result.items[0].url
@@ -70,18 +89,14 @@ const fetchers = {
     }
     finally
     {
-      return {
-        fetchedFrom: undefined,
-        artworkUrl: undefined,
-        joinUrl: undefined
-      };
+      return (returnValue ? returnValue : {});
     }
   }
 }
 
 // Fetches both artwork and join URL if possible,
 // Returns object containing artworkUrl, artworkFrom, joinUrl and joinFrom
-async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, artist, album, song)
+async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, metadata)
 {
   let artworkUrl, artworkFrom, joinUrl, joinFrom;
   let results = [];
@@ -89,7 +104,7 @@ async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, ar
   // First try fetching artwork URL using preferred provider
   if (preferredArtworkProvider in fetchers)
   {
-    results[preferredArtworkProvider] = await fetchers[preferredArtworkProvider](artist, album, song);
+    results[preferredArtworkProvider] = await fetchers[preferredArtworkProvider](metadata);
     if(results[preferredArtworkProvider].artworkUrl)
     {
       artworkUrl = results[preferredArtworkProvider].artworkUrl;
@@ -97,16 +112,18 @@ async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, ar
     }
   }
 
-  // Next try fetching join URL
-  if (preferredArtworkProvider !== preferredJoinProvider
-      && preferredJoinProvider in fetchers)
+  // Next try fetching join URL from preferred provider
+  if (preferredJoinProvider in fetchers
+      && preferredArtworkProvider !== preferredJoinProvider)
   {
-    results[preferredJoinProvider] = await fetchers[preferredJoinProvider](artist, album, song);
-    if (results[preferredJoinProvider].joinUrl)
-    {
-      joinUrl = results[preferredJoinProvider].joinUrl;
-      joinFrom = results[preferredJoinProvider].fetchedFrom;
-    }
+    results[preferredJoinProvider] = await fetchers[preferredJoinProvider](metadata);
+  }
+  // Set it separately, in case both preferred providers are the same
+  if (preferredJoinProvider in results
+      && results[preferredJoinProvider].joinUrl)
+  {
+    joinUrl = results[preferredJoinProvider].joinUrl;
+    joinFrom = results[preferredJoinProvider].fetchedFrom;
   }
 
   // Try using preferred join provider as a backup artwork provider and vice versa
@@ -127,7 +144,7 @@ async function combinedFetch(preferredArtworkProvider, preferredJoinProvider, ar
   {
     if (!(availableProviderNames[ii] in results))
     {
-      results[availableProviderNames[ii]] = await fetchers[availableProviderNames[ii]](artist, album, song);
+      results[availableProviderNames[ii]] = await fetchers[availableProviderNames[ii]](metadata);
       if(!artworkUrl && results[availableProviderNames[ii]].artworkUrl)
       {
         artworkUrl = results[availableProviderNames[ii]].artworkUrl;
@@ -159,37 +176,18 @@ module.exports = async (status) => {
   } // else
   const { meta } = status.information.category;
 
-
-  // Find correct artist name to display,
-  //  then append one character since Discord^{citation needed} doesn't like
-  //  when it's only one character.
-  let display_artist = undefined;
-  if (meta.albumartist)
-  {
-    display_artist = meta.albumartist + " ";
-  }
-  else if (meta.artist)
-  {
-    display_artist = meta.artist + " ";
-  }
-
   // Fetch artwork and join URLs
-  let artwork, fetched, joinUrl, joinLabel;
+  let artwork = config.rpc.largeIcon, fetched = "Nowhere", joinUrl, joinLabel;
 
-  if (meta.title !== undefined && display_artist !== undefined)
   {
-    const fetchResult = await combinedFetch(config.rpc.whereToFetchOnline, config.rpc.changeButtonProvider, display_artist, meta.album, meta.title);
+    const fetchResult
+      = await combinedFetch(config.rpc.whereToFetchOnline, config.rpc.changeButtonProvider, meta);
 
     // Set artwork URL if present
     if(fetchResult.artworkUrl != undefined)
     {
       artwork = fetchResult.artworkUrl;
       fetched = fetchResult.artworkFrom;
-    }
-    else
-    {
-      artwork = config.rpc.largeIcon;
-      fetched = "Nowhere";
     }
 
     // Set join URL if present
@@ -199,21 +197,13 @@ module.exports = async (status) => {
       joinLabel = `Listen on ${fetchResult.joinFrom}`;
     }
   }
-  else
-  {
-    artwork = config.rpc.largeIcon;
-    fetched = "Nowhere";
-  }
 
   if (config.debug === 'true') {
     console.log(artwork);
     console.log(status.state);
     console.log(fetched);
-    console.log(meta.title);
-    console.log(meta.artist, artist);
-    console.log(display_artist)
-    console.log(meta.albumartist)
-    console.log(status.stats.decodedvideo)
+    console.log(meta);
+    console.log(status.stats.decodedvideo);
   }
 
   // Large image hover text
@@ -230,7 +220,7 @@ module.exports = async (status) => {
     largeImageTextIs = `Artwork fetched from ${fetched}`;
   }
 
-  output = {
+  let output = {
     // Shows file thats playing.. well most of the time
     details: meta.title || meta.filename || "Playing something..",
     largeImageText: largeImageTextIs,
@@ -250,16 +240,26 @@ module.exports = async (status) => {
       }
     ];
   }
-  
-  // if video
-  if(status.stats.decodedvideo > 0) {
-    // if youtube video
-    if (meta['YouTube Start Time'] !== undefined) {
+
+  // Find correct artist name to display,
+  //  then append one character since Discord state must be at least two characters long
+  let display_artist = (meta.albumartist
+                        ? meta.albumartist + " "
+                        : (meta.artist
+                           ? meta.artist + " "
+                           : undefined));
+
+  if(status.stats.decodedvideo > 0)
+  { // if video
+    if (meta['YouTube Start Time'] !== undefined)
+    { // if youtube video
       output.largeImageKey = 'youtube';
       output.largeImageText = meta.url;
     }
-    // if a tv show
-    if (meta.showName) output.details = meta.showName;
+    if (meta.showName)
+    { // if a tv show
+      output.details = meta.showName;
+    }
     if (meta.episodeNumber) {
       output.state = `Episode ${meta.episodeNumber}`;
       if (meta.seasonNumber) {
@@ -270,10 +270,14 @@ module.exports = async (status) => {
     } else {
       output.state = `${(status.date || '')} Video`;
     }
-  } else if (meta.now_playing) {
-    // if a stream
+  }
+  else if (meta.now_playing)
+  { // if a stream
     output.state = meta.now_playing || "Stream";
-  } else if (display_artist) {
+  }
+  else if (display_artist)
+  { // if a song
+
     // Add artist to the state
     output.state = display_artist;
 
@@ -281,19 +285,22 @@ module.exports = async (status) => {
     if (meta.album)
       output.state += ` - ${meta.album}`;
 
-    //Trim the state if too longdisplay_artist
+    // Trim the state if too long
     if (output.state.length > 128)
     {
       console.warn("The string ('" + output.state + "') is too long for discord :/");
       output.state = output.state.substring(0, 128);
     }
 
-    // display track #
-    if (meta.track_number && meta.track_total && config.rpc.displayTrackNumber) {
+    // Display track number
+    if (meta.track_number && meta.track_total && config.rpc.displayTrackNumber)
+    {
       output.partySize = parseInt(meta.track_number, 10);
       output.partyMax = parseInt(meta.track_total, 10);
     }
-  } else {
+  }
+  else
+  {
     output.state = status.state;
   }
   const end = Math.floor(Date.now() / 1000 + ((status.length - status.time) / status.rate));

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -12,6 +12,67 @@ const yt = require("ytsr");
 
 
 
+// These functions provide uniform inteface for simple access to the APIs
+// They take artist album and song as the arguments and return object containing
+//   fetchedFrom, artworkUrl and joinUrl (or undefined where not applicable).
+const fetchers = {
+  "apple": async (artist, album, song) => {
+    const params = {
+      media: "music",
+      term: `${artist} ${song}`,
+    };
+    const result = await axios.get("https://itunes.apple.com/search", {
+      params,
+    });
+    if (result.data.resultCount == 0 || result.data.results[0] === undefined)
+    {
+      return {
+        fetchedFrom: undefined,
+        artworkUrl: undefined,
+        joinUrl: undefined
+      };
+    }
+    else
+    {
+      return {
+        fetchedFrom: "Apple Music",
+        artworkUrl: result.data.results[0].artworkUrl100,
+        joinUrl: result.data.results[0].trackViewUrl
+      };
+    }
+  },
+  "spotify": async (artist, album, song) => {
+    let result = await albumArt(artist, {album}).then((data) => data);
+    // TODO: how is failure indicated in the albumArt library?
+    return {
+      fetchedFrom: "Spotify",
+      artworkUrl: result,
+      joinUrl: undefined
+    };
+  },
+  "youtube": async (artist, album, song) => {
+    // This just crashes most of the time, no clue why
+    const result = await yt(encodeURI(`${artist} ${song}`), { limit: 1 });
+    if (result.items.length == 0)
+    {
+      return {
+        fetchedFrom: undefined,
+        artworkUrl: undefined,
+        joinUrl: undefined
+      };
+    }
+    else
+    {
+      return {
+        fetchedFrom: "Youtube",
+        artworkUrl: undefined,
+        joinUrl: result.items[0].url
+      };
+    }
+  }
+}
+
+
 module.exports = async (status) => {
   // if playback is stopped
   if (status.state === 'stopped') {
@@ -26,18 +87,9 @@ module.exports = async (status) => {
   const { meta } = status.information.category;
 
 
-
-
-  const fetchArtworkApple = async (searchQuery) => {
-    const params = {
-      media: "music",
-      term: searchQuery,
-    };
-    return axios.get("https://itunes.apple.com/search", {
-      params,
-    });
-  };
-
+  // Find correct artist name to display,
+  //  then append one character since Discord^{citation needed} doesn't like
+  //  when it's only one character.
   let display_artist = undefined;
   if (meta.albumartist)
   {
@@ -47,40 +99,51 @@ module.exports = async (status) => {
   {
     display_artist = meta.artist + " ";
   }
-  
-  const options = {
-    album: meta.album
-  }
 
-  var appleresponse = await fetchArtworkApple(`${meta.title} ${display_artist}`);
-  
-  if (!(meta.title === undefined || display_artist === undefined))
+  // Fetch artwork and join URLs
+  let artwork, fetched, joinUrl, joinLabel;
+
+  if (meta.title !== undefined && display_artist !== undefined)
   {
-    if (config.rpc.whereToFetchOnline === 'apple')
+    // Initial fetch for artwork
+    const fetchResult1 = await fetchers[config.rpc.whereToFetchOnline](display_artist, meta.album, meta.title);
+    if(fetchResult1.artworkUrl != undefined)
     {
-      if (appleresponse.length == 0 || appleresponse.data.results[0] === undefined) {
-        try{
-          var testartwork = await albumArt(display_artist, options).then((data) => data);
-          var artwork = testartwork;
-          var fetched = "Spotify";
-          var enableYoutubeButton = true;
-        } catch (err) {
-          var artwork = config.rpc.largeIcon;
-          var fetched = "Nowhere";
-          var enableYoutubeButton = "true";
-          console.log(err)
-        }
-      } else {
-        var artwork = appleresponse.data.results[0].artworkUrl100;
-        var fetched = "Apple";
+      artwork = fetchResult1.artworkUrl;
+      fetched = fetchResult1.fetchedFrom;
+    }
+    else
+    {
+      artwork = config.rpc.largeIcon;
+      fetched = "Nowhere";
+    }
+
+    // If desired artwork and join URL providers are the same,
+    //  use existing fetch, otherwise fetch the join URL
+    if (config.rpc.whereToFetchOnline === config.rpc.changeButtonProvider)
+    {
+      if (fetchResult1.joinUrl)
+      {
+        joinUrl = fetchResult1.joinUrl;
+        joinLabel = `Listen on ${fetchResult1.fetchedFrom}`;
       }
-    } else if (config.rpc.whereToFetchOnline === 'spotify') {
-      var artwork = await albumArt(display_artist, options).then((data) => data);
-      var fetched = "Spotify";
+    }
+    else
+    {
+      const fetchResult2 = await fetchers[config.rpc.changeButtonProvider](display_artist, meta.album, meta.title);
+      if(fetchResult2.joinUrl)
+      {
+        joinUrl = fetchResult2.joinUrl;
+        joinLabel = `Listen on ${fetchResult2.fetchedFrom}`;
+      }
     }
   }
+  else
+  {
+    artwork = config.rpc.largeIcon;
+    fetched = "Nowhere";
+  }
 
-  
   if (config.debug === 'true') {
     console.log(artwork);
     console.log(status.state);
@@ -92,92 +155,41 @@ module.exports = async (status) => {
     console.log(status.stats.decodedvideo)
   }
 
-  if(config.rpc.changeButtonProvider === "youtube"){
-    var enableYoutubeButton = "true";
-  } 
-
-  if (display_artist === undefined){
-    var artwork = config.rpc.largeIcon;
-    var fetched = "Nowhere";
-    var enableYoutubeButton = "true";
-  } else if (meta.title === undefined) {
-    var artwork = config.rpc.largeIcon;
-    var fetched = "Nowhere";
-    var enableYoutubeButton = "true";
-  }
-
-
-  // Really not much
+  // Large image hover text
+  let largeImageTextIs;
   if (config.rpc.largeImageText === "artist"){
-    var largeImageTextIs = decodeURI(artist);
+    largeImageTextIs = decodeURI(artist);
   } else if (config.rpc.largeImageText === "album") {
-    var largeImageTextIs = meta.album;
+    largeImageTextIs = meta.album;
   } else if (config.rpc.largeImageText === "volume") {
-    var largeImageTextIs = `Volume: ${Math.round(status.volume / 2.56)}%`;
+    largeImageTextIs = `Volume: ${Math.round(status.volume / 2.56)}%`;
   } else if (config.rpc.largeImageText === "title") {
-    var largeImageTextIs = meta.title;
+    largeImageTextIs = meta.title;
   } else if (config.rpc.largeImageText === "fetched") {
-    var largeImageTextIs = `Artwork fetched from ${fetched}`;
-  }
-  
-// Checks if youtube button is turned on in the config, then searches the names on youtube
-  if(enableYoutubeButton){
-    if (meta.title === undefined) {
-      const search = await yt(`${meta.filename}`, { limit: 1 });
-      const resultunjson = JSON.stringify(search.items);
-      const result = JSON.parse(resultunjson);
-      if (result[0] === undefined) {
-        var url = "https://videolan.com/vlc";
-        var label = "Visit VLC today!";
-      } else {
-        var url = result[0].url;
-        var label = "Listen on Youtube"; 
-      }
-      
-    } else if (display_artist === undefined) {
-      const search = await yt(`${meta.filename}`, { limit: 1 });
-      const resultunjson = JSON.stringify(search.items);
-      const result = JSON.parse(resultunjson);
-      if (result[0] === undefined) {
-        var url = "https://videolan.com/vlc";
-        var label = "Visit VLC today!";
-      } else {
-        var url = result[0].url;
-        var label = "Listen on Youtube";
-      }
-    } else {
-      const search = await yt(`${meta.title} ${display_artist}`, { limit: 1 });
-      const resultunjson = JSON.stringify(search.items);
-      const result = JSON.parse(resultunjson);
-      var url = result[0].url;
-      var label = "Listen on Youtube";
-    }
-    
-
-  } else {
-    var url = appleresponse.data.results[0].trackViewUrl;
-    
-    var label = "Listen on Apple Music";
+    largeImageTextIs = `Artwork fetched from ${fetched}`;
   }
 
-
-  const output = {
+  output = {
     // Shows file thats playing.. well most of the time
     details: meta.title || meta.filename || "Playing something..",
-    // Line 99 - 109 determines this
     largeImageText: largeImageTextIs,
     // Sets album art depending on whats set in the file, or if album art cannot be found
     largeImageKey: artwork || "https://i.pinimg.com/originals/67/f6/cb/67f6cb14f862297e3c145014cdd6b635.jpg",
     smallImageKey: status.state,
     smallImageText: `Volume: ${Math.round(status.volume / 2.56)}%`,
     instance: true,
-    buttons: [
-        {
-          label: label,
-          url: url,
-        },
-      ]
   };
+  
+  if (joinUrl && joinLabel)
+  {
+    output.buttons = [
+      {
+        label: joinLabel,
+        url: joinUrl
+      }
+    ];
+  }
+  
   // if video
   if(status.stats.decodedvideo > 0) {
     // if youtube video

--- a/src/rpc/format.js
+++ b/src/rpc/format.js
@@ -29,7 +29,7 @@ const fetchers = {
       });
       if (result.data.resultCount > 0 && result.data.results[0] !== undefined)
       {
-        returnValue = {
+        return {
           fetchedFrom: "Apple Music",
           artworkUrl: result.data.results[0].artworkUrl100,
           joinUrl: result.data.results[0].trackViewUrl

